File: src/codegen/env.py
from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path
from typing import Mapping, Sequence

from codegen.scope import ScopeStore


@dataclass
class RunContext:
    invoke_cwd: Path
    targets: Sequence[Path]
    file_path: Path
    origin_file_path: Path        # temp file path containing pre-processing snapshot
    origin_block_path: Path       # temp file path containing original block content


def build_env(
    cfg_extra_env: Mapping[str, str],
    scope: ScopeStore,
    ctx: RunContext,
) -> dict[str, str]:
    """Assemble the subprocess environment for a block execution (§4).

    Order: os.environ → cfg.extra_env → runtime codegen vars.
    Later entries win.
    """
    env = dict(os.environ)

    for k, v in cfg_extra_env.items():
        env[k] = v

    env["CODEGEN_GLOBAL"] = str(scope.global_path)
    env["CODEGEN_FILE"] = str(scope.file_path)
    env["CODEGEN_BLOCK"] = str(scope.block_path)

    env["CODEGEN_ORIGIN_FILE"] = str(ctx.origin_file_path)
    env["CODEGEN_ORIGIN_BLOCK"] = str(ctx.origin_block_path)

    env["CODEGEN_INVOKE_CWD"] = str(ctx.invoke_cwd)
    env["CODEGEN_TARGETS"] = "\n".join(str(p) for p in ctx.targets)
    env["CODEGEN_FILE_PATH"] = str(ctx.file_path)

    return env

File: cpp/src/env.cpp
#include "env.hpp"
#include "utils.hpp"

#ifdef _WIN32
extern char** _environ;
#define environ _environ
#else
extern char** environ;
#endif

std::map<std::string, std::string> build_env(
    const std::map<std::string, std::string>& cfg_extra_env,
    const ScopeStore& scope,
    const RunContext& ctx)
{
    std::map<std::string, std::string> env;

    // Start from current process environment
    for (char** ep = environ; ep && *ep; ep++) {
        std::string entry(*ep);
        auto eq = entry.find('=');
        if (eq == std::string::npos) continue;
        env[entry.substr(0, eq)] = entry.substr(eq + 1);
    }

    // cfg.extra_env
    for (auto& [k, v] : cfg_extra_env)
        env[k] = v;

    // Scope paths
    env["CODEGEN_GLOBAL"] = scope.global_path().string();
    env["CODEGEN_FILE"]   = scope.file_path().string();
    env["CODEGEN_BLOCK"]  = scope.block_path().string();

    // Origin snapshots
    env["CODEGEN_ORIGIN_FILE"]  = ctx.origin_file_path.string();
    env["CODEGEN_ORIGIN_BLOCK"] = ctx.origin_block_path.string();

    // Path/location indicators
    env["CODEGEN_INVOKE_CWD"] = ctx.invoke_cwd.string();

    std::string targets_str;
    for (size_t i = 0; i < ctx.targets.size(); i++) {
        if (i > 0) targets_str += '\n';
        targets_str += ctx.targets[i].string();
    }
    env["CODEGEN_TARGETS"]   = targets_str;
    env["CODEGEN_FILE_PATH"] = ctx.file_path.string();

    return env;
}
